### PR TITLE
Refactor: CType Utility and Split Metadata and Schema

### LINKED ISSUE #200: Have notification on starting a transfer

### PR DESCRIPTION
**Pull request by Dudleyneedham**
## fixes https://github.com/KILTprotocol/ticket/issues/15

Removed CTypeUtils functions from the SDK

- [x] fromInputModel
- [x] getClaimInputModel
- [x] getCTypeInputModel
- [x] Checked the code in the demo client works

Added them into the demo client.

## How to test:

Connect to the chain with the changes, did some test and went through the various Ctype and claims functionality. 

You will need to use the Link function and pull this version of the SDK https://github.com/KILTprotocol/sdk-js/pull/211 in order to test.

Then, periodically test the ctype and claims. 

Along with the dev tools, though I think they aren't working correctly in the first place.
 
## Checklist:

- [x] I have verified that the code works
- [x] I have verified that the code is easy to understand
  - [ ] If not, I have left a well-balanced amount of inline comments
- [x] I have [left the code in a better state](https://deviq.com/boy-scout-rule/)
- [x] I have documented the changes (where applicable)

### CODE REVIEW

**Comment by Dudleyneedham** [comment]
Need to clean up the types used within the functions.

File: src/containers/CtypeCreate/CtypeCreate.tsx [hunk 1]
```diff
@@ -137,6 +138,10 @@ class CTypeCreate extends React.Component<Props, State> {
       isValid,
     })
   }
+
+  private inputModel(ctypeInput: ICTypeInput): sdk.CType {
```
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply src/containers/CtypeCreate/CtypeCreate.tsx]: I have updated it accordingly.

File: src/containers/Tasks/SubmitLegitimations/SubmitLegitimations.tsx [hunk 2]
```diff
@@ -190,6 +192,9 @@ class SubmitLegitimations extends React.Component<Props, State> {
       }
     })
   }
+  private claimInputModel(cType: sdk.ICType, lang?: string): IClaimInput {
```
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply src/containers/Tasks/SubmitLegitimations/SubmitLegitimations.tsx]: I have updated it accordingly.

File: src/services/CtypeUtils.ts [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+/**
+ * Create the CTYPE model from a CTYPE input model (used in CTYPE editing components).
+ * This is necessary because component editors rely on editing arrays of properties instead of
+ * arbitrary properties of an object. Additionally the default language translations are integrated
+ * into the input model and need to be separated for the CTYPE model.
+ * This is the reverse function of CType.getCTypeInputModel(...).
+ * @returns The CTYPE for the input model.
+ */
+
+class CTypeUtils {
```
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply src/services/CtypeUtils.ts]: I have updated it accordingly.

File: src/components/MyClaimCreateView/MyClaimCreateView.tsx [hunk 4]
```diff
@@ -166,6 +164,10 @@ class MyClaimCreateView extends Component<Props, State> {
   private handleNameChange(e: React.ChangeEvent<HTMLInputElement>) {
     this.setState({ name: e.target.value })
   }
+
+  private claimInputModel(cType: sdk.ICType, lang?: string): IClaimInput {
```
[the remark this answers is not in the document]
Comment (Dudleyneedham) [reply src/components/MyClaimCreateView/MyClaimCreateView.tsx]: Removed all the private functions.

**Comment by Dudleyneedham** [comment]
I have added the test into the file, but they don't work.

**Comment by tjwelde** [comment]
Please put the ctypeutils in the `utils`folder: https://github.com/KILTprotocol/demo-client/tree/dn-demo-200-ctype-utils-refac-objec/src/utils

**Comment by Dudleyneedham** [comment]
I made changes accordingly, additionally.

I was testing the changes in splitting the ctype metadata and schema. 

These were closely related and I felt I needed to be added at the same time. They work, but I had to take some checks off. I am trying to think of a way to fix the any.

File: src/components/CTypePresentation/CTypePresentation.tsx [hunk 5]
```diff
@@ -91,7 +91,7 @@ class CTypePresentation extends React.Component<Props, State> {
               size={size || DEFAULT_SIZE}
               theme="polkadot"
             />
-            {this.getLabel(cType.cType.metadata.title.default)}
+            {this.getLabel(cType.cType.schema.$id)}
```
Comment (Dudleyneedham) [inline src/components/CTypePresentation/CTypePresentation.tsx]: A quick fix, overall I would need to change the state to match the new ctype. Unfortunately, it is a lot of work and seems out of the scope of the ticket.

File: src/components/CtypeDetailView/CtypeDetailView.tsx [hunk 6]
```diff
@@ -41,7 +41,7 @@ class CtypeDetailView extends React.Component<Props, State> {
             <div className="attributes">
               <div>
                 <label>Title</label>
-                <div>{cType.cType.metadata.title.default}</div>
+                <div>{cType.cType.schema.$id}</div>
```
Comment (Dudleyneedham) [inline src/components/CtypeDetailView/CtypeDetailView.tsx]: see comment above

File: src/components/DevTools/DevTools.ctypes.tsx [hunk 7]
```diff
@@ -35,7 +35,7 @@ class BsCType {
     return cType
       .store(ownerIdentity)
       .then((value: any) => {
-        const cTypeWrapper: ICType = {
+        const cTypeWrapper: any = {
```
Comment (Dudleyneedham) [inline src/components/DevTools/DevTools.ctypes.tsx]: The only problem I am currently facing, I am able to find a way around this without changing the whole state. As we take the sdk.CType as the state. Which has been changed by removing the metadata. 

It would require me to take the demo client apart and feels unnecessary to complete the ticket. 

I tested the functionality and if someone uses the utilities to create cTypes, I believe it encounters a problem.

File: src/components/MyClaimCreateView/MyClaimCreateView.tsx [hunk 8]
```diff
@@ -102,9 +101,7 @@ class MyClaimCreateView extends Component<Props, State> {
               </div>
             </div>
             <SchemaEditor
-              schema={
-                sdk.CTypeUtils.getClaimInputModel(cType!) as common.Schema
-              }
+              schema={getClaimInputModel(cType!) as common.Schema}
```
Comment (Dudleyneedham) [inline src/components/MyClaimCreateView/MyClaimCreateView.tsx]: I have altered the getClaimInputModel to match the new CType class inside the demo client. Feels a little hacky, plus I had to remove some information. 

I will bring this up and hopefully see if there is a better way of doing it.

File: src/components/SelectCTypes/SelectCTypes.tsx [hunk 9]
```diff
@@ -127,7 +127,7 @@ class SelectCTypes extends React.Component<Props, State> {
     return {
       baseValue: `${cType.cType.hash}`,
       label: <CTypePresentation cTypeHash={cType.cType.hash} />,
-      value: `${cType.cType.metadata.title.default} ${cType.cType.hash}`,
+      value: `${cType.cType.schema.$id} ${cType.cType.hash}`,
```
Comment (Dudleyneedham) [inline src/components/SelectCTypes/SelectCTypes.tsx]: Same as above

File: src/containers/CtypeCreate/CtypeCreate.tsx [hunk 10]
```diff
@@ -83,18 +84,22 @@ class CTypeCreate extends React.Component<Props, State> {
           blockUi.updateMessage(
             `CTYPE stored on blockchain,\nnow registering CTYPE`
           )
-          const cTypeWrapper: ICType = {
-            cType,
+          const cTypeWrapper: CTypeMetadata = {
```
Comment (Dudleyneedham) [inline src/containers/CtypeCreate/CtypeCreate.tsx]: I had to create a new type interface in order to match what is taken in the database.

I replicated the data after testing with my changes and then alongside what the demo client currently produces. I was able to get the information to match and not make so many changes to the backend. Unfortunately, there was a minor change due to a test and console log.

I have tried to match the state as seen above with `metadata`. Ideally, I would have liked to change a lot more. After attempting this I realised it was wasting too much time and felt unnecessary for the ticket.

File: src/services/CtypeRepository.ts [hunk 11]
```diff
@@ -45,7 +45,7 @@ class CTypeRepository {
       })
   }
 
-  public static async register(cType: ICType): Promise<Response> {
+  public static async register(cType: CTypeMetadata): Promise<Response> {
```
Comment (Dudleyneedham) [inline src/services/CtypeRepository.ts]: I had to change the type to match the input. Unfortunately, this is the problem I am facing with the BScType within the dev tools. It takes the sdk.Ctype, not the newly created CTypeMetadata.

File: src/types/Ctype.ts [hunk 12]
```diff
@@ -1,12 +1,46 @@
...
+export interface IClaimInput {
+  $id: string
+  $schema: string
+  properties: object
+  required: string[]
+  title: string
+  description?: string
+  type: string
+  owner?: string
+}
```
Comment (Dudleyneedham) [inline src/types/Ctype.ts]: I had taken all the information from the sdk CTypeUtilites and added it here. 

Additionally, I created a new typeface to take the combined schema and metadata to match what we previously had. 

It was a workaround in order not to take up too much time.

File: src/utils/CtypeUtils/index.ts [hunk 13]
```diff
@@ -0,0 +1 @@
+export * from './CtypeUtils'
```
Comment (Dudleyneedham) [inline src/utils/CtypeUtils/index.ts]: Created the CTypeUtils as suggested.

File: src/utils/CtypeUtils/CtypeUtils.ts [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+export const getClaimInputModel = (
+  ctype: sdk.CType,
+  lang?: string
+): IClaimInput => {
+  // create clone
+  const result = JSON.parse(JSON.stringify(ctype.schema))
+  result.required = []
+
+  return result
+}
```
Comment (Dudleyneedham) [inline src/utils/CtypeUtils/CtypeUtils.ts]: I have modified this file to take all the information necessary for the split ctype. 

This is the reason I merged these pull request and did this on one branch. 

The changes had not been implemented on the demo client and they needed to be.

File: src/types/Ctype.ts [hunk 15]
```diff
@@ -1,10 +1,17 @@
...
-  metaData: {
-    author: string
-  }
+}
+
+export interface CTypeMetadataChain {
+  cType: sdk.CType
+  metaData: ICTypeMetadata
+}
+export interface CTypeMetadata {
```
Comment (tjwelde) [inline src/types/Ctype.ts]: This reads, as if this only contains ctype metadata and not cype AND metadata

File: src/services/DidService.ts [hunk 16]
```diff
@@ -23,7 +23,7 @@ export class DidService {
       ContactRepository.URL
     }/${myIdentity.identity.address}`
     const did = sdk.Did.fromIdentity(myIdentity.identity, documentStore)
-    const didDocument = did.getDefaultDocument(`${MessageRepository.URL}`)
+    const didDocument = did.createDefaultDidDocument(`${MessageRepository.URL}`)
```
Comment (tjwelde) [inline src/services/DidService.ts]: Oh, was this overlooked?

File: src/components/CtypeListView/CtypeListView.tsx [hunk 17]
```diff
@@ -64,7 +64,7 @@ class CtypeListView extends React.Component<Props, State> {
                       right={true}
                     />
                     <ContactPresentation
-                      address={cType.metaData.author}
+                      address={cType.cType.owner!}
```
Comment (tjwelde) [inline src/components/CtypeListView/CtypeListView.tsx]: yes!

File: src/types/Ctype.ts [hunk 18]
```diff
@@ -1,10 +1,17 @@
...
 
-export interface ICType {
+export interface ICType extends sdk.ICTypeMetadata {
   cType: sdk.ICType
-  metaData: {
-    author: string
-  }
+}
+
+export interface CTypeMetadataChain {
```
Comment (tjwelde) [inline src/types/Ctype.ts]: Same as below. Why do we even need this interface?

File: src/types/Ctype.ts [hunk 19]
```diff
@@ -1,10 +1,17 @@
 import * as sdk from '@kiltprotocol/sdk-js'
+import ICTypeMetadata, { IMetadata } from '@kiltprotocol/sdk-js/build/types/CTypeMetedata'
```
Comment (tjwelde) [inline src/types/Ctype.ts]: please try to import everything from the index

File: src/types/Ctype.ts [hunk 20]
```diff
@@ -25,20 +33,19 @@ export interface IClaimInput {
...
-  public metaData: {
-    author: string
-  }
   public readonly cType: sdk.ICType
-
+  public metadata: IMetadata
+  public ctypeHash: sdk.ICType['hash']
   public getPropertyTitle(propertyName: string) {
-    return this.cType.metadata.properties[propertyName].title.default
+    return propertyName // Need to fix. I need to take the value of the key.
```
Comment (tjwelde) [inline src/types/Ctype.ts]: This needs to be fixed, before we merge

File: src/types/Ctype.ts [hunk 21]
```diff
@@ -25,20 +33,19 @@ export interface IClaimInput {
...
     const newCtype = Object.create(CType.prototype)
     return Object.assign(newCtype, obj)
   }
 
-  public metaData: {
-    author: string
-  }
   public readonly cType: sdk.ICType
-
+  public metadata: IMetadata
```
Comment (tjwelde) [inline src/types/Ctype.ts]: You might be able to use `ICTypeMetadata['metadata'] ` to shorten your import

File: src/components/CTypePresentation/CTypePresentation.tsx (hunk 5 above)
Comment (tjwelde) [inline src/components/CTypePresentation/CTypePresentation.tsx]: AFAICS, we have access to the metadata here, so lets just use it

File: src/components/CtypeDetailView/CtypeDetailView.tsx (hunk 6 above)
Comment (tjwelde) [reply src/components/CtypeDetailView/CtypeDetailView.tsx]: since `ctype` is of type `CTypeMetadata`, we should be able to get the title from the metadata

File: src/components/SelectCTypes/SelectCTypes.tsx (hunk 9 above)
Comment (tjwelde) [reply src/components/SelectCTypes/SelectCTypes.tsx]: since `ctype` is of type `CTypeMetadata`, we should be able to get the title from the metadata

File: src/utils/CtypeUtils/CtypeUtils.ts [hunk 22]
```diff
@@ -90,19 +108,12 @@ export const getCTypeInputModel = (ctype: sdk.CType): ICTypeInput => {
  * @returns {any} The claim input model
  */
 export const getClaimInputModel = (
-  ctype: sdk.ICType,
+  ctype: sdk.CType,
```
Comment (tjwelde) [inline src/utils/CtypeUtils/CtypeUtils.ts]: Why the change from the Interface to the class?

File: src/types/Ctype.ts [hunk 23]
```diff
@@ -25,20 +33,19 @@ export interface IClaimInput {
   title: string
   description?: string
   type: string
+  owner?: string
 }
 
 export class CType implements ICType {
-  public static fromObject(obj: ICType): CType {
+  public static fromObject(obj: CTypeMetadata): CType {
```
Comment (tjwelde) [inline src/types/Ctype.ts]: Can we get rid of this whole class, or would it be too much?

File: src/types/Ctype.ts (hunk 15 above)
Comment (LeonFLK) [reply src/types/Ctype.ts]: renamed this interface to `CTypeWithMetadata`

File: src/services/DidService.ts (hunk 16 above)
Comment (LeonFLK) [reply src/services/DidService.ts]: Yes this was overlooked i guess.

File: src/types/Ctype.ts (hunk 18 above)
Comment (LeonFLK) [reply src/types/Ctype.ts]: Removed

File: src/types/Ctype.ts (hunk 19 above)
Comment (LeonFLK) [reply src/types/Ctype.ts]: Yeah, I should check for this auto import issue more often

File: src/types/Ctype.ts (hunk 21 above)
Comment (LeonFLK) [reply src/types/Ctype.ts]: Yes

File: src/types/Ctype.ts (hunk 20 above)
Comment (LeonFLK) [reply src/types/Ctype.ts]: I reverted it to the previous state, might not work as we want it to be, I would need your input.

File: src/components/CTypePresentation/CTypePresentation.tsx (hunk 5 above)
Comment (LeonFLK) [reply src/components/CTypePresentation/CTypePresentation.tsx]: yes we do have access.

File: src/utils/CtypeUtils/CtypeUtils.ts (hunk 22 above)
Comment (LeonFLK) [reply src/utils/CtypeUtils/CtypeUtils.ts]: this had to be a mistake, I reverted it.

File: src/types/Ctype.ts (hunk 23 above)
Comment (LeonFLK) [reply src/types/Ctype.ts]: Yes this is possible, it's functionality is only used in a few places, I removed it for now and used our old method of acquiring the property title and extended it with an additional check.

```
cType && cType.metadata.properties[propertyName].title.default ?
      cType.metadata.properties[propertyName].title.default : propertyName
```

File: src/types/Ctype.ts (hunk 20 above)
Comment (LeonFLK) [reply src/types/Ctype.ts]: removed it and moved the logic to the original calling functions `getCtypePropertyTitle` in 

src/components/SelectAttestedClaim/SelectAttestedClaim.tsx
src/components/AttestedClaimVerificationView/AttestedClaimVerificationView.tsx

with an additional check whether the property exists on the metadata:

`cType && cType.metadata.properties[propertyName].title.default ?
      cType.metadata.properties[propertyName].title.default : propertyName`

File: src/components/AttestedClaimVerificationView/AttestedClaimVerificationView.tsx [hunk 24]
```diff
@@ -88,7 +87,9 @@ class AttestedClaimVerificationView extends React.Component<Props, State> {
     return (
       <div className="attributes">
         {propertyNames.map((propertyName: string) => {
-          const propertyTitle = this.getCtypePropertyTitle(propertyName)
+          const propertyTitle = this.props.cType
+            ? getCtypePropertyTitle(propertyName, this.props.cType)
+            : propertyName
```
Comment (LeonFLK) [inline src/components/AttestedClaimVerificationView/AttestedClaimVerificationView.tsx]: Moved existence checks to the calling function, getCtypePropertyTitle now always has to take a ctype.

File: src/components/DevTools/DevTools.attestations.tsx [hunk 25]
```diff
@@ -240,9 +241,9 @@ class BsAttestation {
     const { attesterKey } = attest
 
     const attesterIdentity: MyIdentity = await BsIdentity.getByKey(attesterKey)
-    const cType: ICType = await BsCType.getByKey(bsClaim.cTypeKey)
+    const cType: CTypeWithMetadata = await BsCType.getByKey(bsClaim.cTypeKey)
 
-    const partialClaim = {
+    const partialClaim: IPartialClaim = {
       cTypeHash: cType.cType.hash as string,
```
Comment (Dudleyneedham) [inline src/components/DevTools/DevTools.attestations.tsx]: Does it need the `as string`?

File: src/types/Ctype.ts [hunk 26]
```diff
@@ -1,10 +1,12 @@
 import * as sdk from '@kiltprotocol/sdk-js'
 
-export interface ICType {
+export interface ICType extends sdk.ICTypeMetadata {
   cType: sdk.ICType
-  metaData: {
-    author: string
-  }
+}
```
Comment (Dudleyneedham) [inline src/types/Ctype.ts]: Glad it is gone.

File: src/utils/CtypeUtils/CtypeUtils.ts [hunk 27]
```diff
@@ -64,23 +95,25 @@ export const getLocalized = (o: any, lang?: string): string => {
...
 
-  Object.entries(ctype.schema.properties as object).forEach(([key, value]) => {
-    result.properties.push({
-      title: getLocalized(ctype.metadata.properties[key].title),
-      $id: key,
-      type: value.type,
-    })
-    result.required.push(key)
-  })
+  Object.entries(ctype.cType.schema.properties as object).forEach(
```
Comment (Dudleyneedham) [inline src/utils/CtypeUtils/CtypeUtils.ts]: Should we remove the properties `as object`?

File: src/types/Ctype.ts [hunk 28]
```diff
@@ -13,8 +15,9 @@ export interface ICTypeInput {
   properties: object[] // TO DO: need to refine what properties are
```
Comment (Dudleyneedham) [inline src/types/Ctype.ts]: I don't think this should be an Array of objects?

File: src/components/CTypePresentation/CTypePresentation.tsx (hunk 5 above)
Comment (tjwelde) [reply src/components/CTypePresentation/CTypePresentation.tsx]: So, since we now have `CTypeWithMetadata` as a type, we can access the metadata again, which is done as of right now

File: src/components/CtypeDetailView/CtypeDetailView.tsx (hunk 6 above)
Comment (tjwelde) [reply src/components/CtypeDetailView/CtypeDetailView.tsx]: This is done as of writing this comment

File: src/components/DevTools/DevTools.ctypes.tsx (hunk 7 above)
Comment (tjwelde) [reply src/components/DevTools/DevTools.ctypes.tsx]: @LeonFLK changed it to `CTypeWithMetadata`

File: src/utils/CtypeUtils/CtypeUtils.ts [hunk 29]
```diff
@@ -97,14 +128,7 @@ export const getClaimInputModel = (
...
   const result = JSON.parse(JSON.stringify(ctype.schema))
-  result.title = getLocalized(ctype.metadata.title, lang)
-  result.description = getLocalized(ctype.metadata.description, lang)
   result.required = []
-  Object.entries(ctype.metadata.properties as object).forEach(
-    ([key, value]) => {
-      result.properties[key].title = getLocalized(value.title, lang)
-      result.required.push(key)
-    }
-  )
```
Comment (tjwelde) [inline src/utils/CtypeUtils/CtypeUtils.ts]: Is this handled in another way now?

File: src/components/SelectCTypes/SelectCTypes.tsx (hunk 9 above)
Comment (tjwelde) [reply src/components/SelectCTypes/SelectCTypes.tsx]: @LeonFLK plz check this

File: src/types/Ctype.ts (hunk 28 above)
Comment (tjwelde) [reply src/types/Ctype.ts]: not part of this ticket

File: src/components/DevTools/DevTools.attestations.tsx (hunk 25 above)
Comment (tjwelde) [reply src/components/DevTools/DevTools.attestations.tsx]: not part of this ticket / PR

File: src/utils/CtypeUtils/CtypeUtils.ts (hunk 27 above)
Comment (tjwelde) [reply src/utils/CtypeUtils/CtypeUtils.ts]: not part of this ticket

**Comment by LeonFLK** [comment]
I looked at the code with Dudley, and we found that `getCTypeInputModel` does not get called at all, except in the utils test file.

Do we want to remove it?

File: src/components/MyClaimCreateView/MyClaimCreateView.tsx [hunk 30]
```diff
@@ -30,7 +30,7 @@ type State = {
   partialClaim: sdk.IPartialClaim
   name: string
   isValid: boolean
-  cType?: sdk.CType
+  cType?: CTypeWithMetadata
```
Comment (LeonFLK) [inline src/components/MyClaimCreateView/MyClaimCreateView.tsx]: The state being an actual `sdk.CType` is only of use when we build the claim, so I changed it to `CTypeWithMetadata` work with the updated g`etClaimInputModel`, which now expects an `CTypeWithMetadata`.

File: src/containers/Tasks/SubmitLegitimations/SubmitLegitimations.tsx [hunk 31]
```diff
@@ -31,7 +31,7 @@ type Props = InjectedSelectProps & SubmitLegitimationsProps
 
 type State = {
   claim: sdk.IPartialClaim
-  cType?: sdk.CType
+  cType?: CTypeWithMetadata
```
Comment (LeonFLK) [inline src/containers/Tasks/SubmitLegitimations/SubmitLegitimations.tsx]: here we don't need a sdk.CType at all.

**Comment by tjwelde** [comment]
Regarding the `getCTypeInputModel`: We might want to remove it, but not in this PR. Please make a ticket
